// @discord/embedded-app-sdk/output/utils/PriceConstants.cjs
'use strict';

exports.CurrencyCodes = void 0;
(function (CurrencyCodes) {
    CurrencyCodes["AED"] = "aed";
    CurrencyCodes["AFN"] = "afn";
    CurrencyCodes["ALL"] = "all";
    CurrencyCodes["AMD"] = "amd";
    CurrencyCodes["ANG"] = "ang";
    CurrencyCodes["AOA"] = "aoa";
    CurrencyCodes["ARS"] = "ars";
    CurrencyCodes["AUD"] = "aud";
    CurrencyCodes["AWG"] = "awg";
    CurrencyCodes["AZN"] = "azn";
    CurrencyCodes["BAM"] = "bam";
    CurrencyCodes["BBD"] = "bbd";
    CurrencyCodes["BDT"] = "bdt";
    CurrencyCodes["BGN"] = "bgn";
    CurrencyCodes["BHD"] = "bhd";
    CurrencyCodes["BIF"] = "bif";
    CurrencyCodes["BMD"] = "bmd";
    CurrencyCodes["BND"] = "bnd";
    CurrencyCodes["BOB"] = "bob";
    CurrencyCodes["BOV"] = "bov";
    CurrencyCodes["BRL"] = "brl";
    CurrencyCodes["BSD"] = "bsd";
    CurrencyCodes["BTN"] = "btn";
    CurrencyCodes["BWP"] = "bwp";
    CurrencyCodes["BYN"] = "byn";
    CurrencyCodes["BYR"] = "byr";
    CurrencyCodes["BZD"] = "bzd";
    CurrencyCodes["CAD"] = "cad";
    CurrencyCodes["CDF"] = "cdf";
    CurrencyCodes["CHE"] = "che";
    CurrencyCodes["CHF"] = "chf";
    CurrencyCodes["CHW"] = "chw";
    CurrencyCodes["CLF"] = "clf";
    CurrencyCodes["CLP"] = "clp";
    CurrencyCodes["CNY"] = "cny";
    CurrencyCodes["COP"] = "cop";
    CurrencyCodes["COU"] = "cou";
    CurrencyCodes["CRC"] = "crc";
    CurrencyCodes["CUC"] = "cuc";
    CurrencyCodes["CUP"] = "cup";
    CurrencyCodes["CVE"] = "cve";
    CurrencyCodes["CZK"] = "czk";
    CurrencyCodes["DJF"] = "djf";
    CurrencyCodes["DKK"] = "dkk";
    CurrencyCodes["DOP"] = "dop";
    CurrencyCodes["DZD"] = "dzd";
    CurrencyCodes["EGP"] = "egp";
    CurrencyCodes["ERN"] = "ern";
    CurrencyCodes["ETB"] = "etb";
    CurrencyCodes["EUR"] = "eur";
    CurrencyCodes["FJD"] = "fjd";
    CurrencyCodes["FKP"] = "fkp";
    CurrencyCodes["GBP"] = "gbp";
    CurrencyCodes["GEL"] = "gel";
    CurrencyCodes["GHS"] = "ghs";
    CurrencyCodes["GIP"] = "gip";
    CurrencyCodes["GMD"] = "gmd";
    CurrencyCodes["GNF"] = "gnf";
    CurrencyCodes["GTQ"] = "gtq";
    CurrencyCodes["GYD"] = "gyd";
    CurrencyCodes["HKD"] = "hkd";
    CurrencyCodes["HNL"] = "hnl";
    CurrencyCodes["HRK"] = "hrk";
    CurrencyCodes["HTG"] = "htg";
    CurrencyCodes["HUF"] = "huf";
    CurrencyCodes["IDR"] = "idr";
    CurrencyCodes["ILS"] = "ils";
    CurrencyCodes["INR"] = "inr";
    CurrencyCodes["IQD"] = "iqd";
    CurrencyCodes["IRR"] = "irr";
    CurrencyCodes["ISK"] = "isk";
    CurrencyCodes["JMD"] = "jmd";
    CurrencyCodes["JOD"] = "jod";
    CurrencyCodes["JPY"] = "jpy";
    CurrencyCodes["KES"] = "kes";
    CurrencyCodes["KGS"] = "kgs";
    CurrencyCodes["KHR"] = "khr";
    CurrencyCodes["KMF"] = "kmf";
    CurrencyCodes["KPW"] = "kpw";
    CurrencyCodes["KRW"] = "krw";
    CurrencyCodes["KWD"] = "kwd";
    CurrencyCodes["KYD"] = "kyd";
    CurrencyCodes["KZT"] = "kzt";
    CurrencyCodes["LAK"] = "lak";
    CurrencyCodes["LBP"] = "lbp";
    CurrencyCodes["LKR"] = "lkr";
    CurrencyCodes["LRD"] = "lrd";
    CurrencyCodes["LSL"] = "lsl";
    CurrencyCodes["LTL"] = "ltl";
    CurrencyCodes["LVL"] = "lvl";
    CurrencyCodes["LYD"] = "lyd";
    CurrencyCodes["MAD"] = "mad";
    CurrencyCodes["MDL"] = "mdl";
    CurrencyCodes["MGA"] = "mga";
    CurrencyCodes["MKD"] = "mkd";
    CurrencyCodes["MMK"] = "mmk";
    CurrencyCodes["MNT"] = "mnt";
    CurrencyCodes["MOP"] = "mop";
    CurrencyCodes["MRO"] = "mro";
    CurrencyCodes["MUR"] = "mur";
    CurrencyCodes["MVR"] = "mvr";
    CurrencyCodes["MWK"] = "mwk";
    CurrencyCodes["MXN"] = "mxn";
    CurrencyCodes["MXV"] = "mxv";
    CurrencyCodes["MYR"] = "myr";
    CurrencyCodes["MZN"] = "mzn";
    CurrencyCodes["NAD"] = "nad";
    CurrencyCodes["NGN"] = "ngn";
    CurrencyCodes["NIO"] = "nio";
    CurrencyCodes["NOK"] = "nok";
    CurrencyCodes["NPR"] = "npr";
    CurrencyCodes["NZD"] = "nzd";
    CurrencyCodes["OMR"] = "omr";
    CurrencyCodes["PAB"] = "pab";
    CurrencyCodes["PEN"] = "pen";
    CurrencyCodes["PGK"] = "pgk";
    CurrencyCodes["PHP"] = "php";
    CurrencyCodes["PKR"] = "pkr";
    CurrencyCodes["PLN"] = "pln";
    CurrencyCodes["PYG"] = "pyg";
    CurrencyCodes["QAR"] = "qar";
    CurrencyCodes["RON"] = "ron";
    CurrencyCodes["RSD"] = "rsd";
    CurrencyCodes["RUB"] = "rub";
    CurrencyCodes["RWF"] = "rwf";
    CurrencyCodes["SAR"] = "sar";
    CurrencyCodes["SBD"] = "sbd";
    CurrencyCodes["SCR"] = "scr";
    CurrencyCodes["SDG"] = "sdg";
    CurrencyCodes["SEK"] = "sek";
    CurrencyCodes["SGD"] = "sgd";
    CurrencyCodes["SHP"] = "shp";
    CurrencyCodes["SLL"] = "sll";
    CurrencyCodes["SOS"] = "sos";
    CurrencyCodes["SRD"] = "srd";
    CurrencyCodes["SSP"] = "ssp";
    CurrencyCodes["STD"] = "std";
    CurrencyCodes["SVC"] = "svc";
    CurrencyCodes["SYP"] = "syp";
    CurrencyCodes["SZL"] = "szl";
    CurrencyCodes["THB"] = "thb";
    CurrencyCodes["TJS"] = "tjs";
    CurrencyCodes["TMT"] = "tmt";
    CurrencyCodes["TND"] = "tnd";
    CurrencyCodes["TOP"] = "top";
    CurrencyCodes["TRY"] = "try";
    CurrencyCodes["TTD"] = "ttd";
    CurrencyCodes["TWD"] = "twd";
    CurrencyCodes["TZS"] = "tzs";
    CurrencyCodes["UAH"] = "uah";
    CurrencyCodes["UGX"] = "ugx";
    CurrencyCodes["USD"] = "usd";
    CurrencyCodes["USN"] = "usn";
    CurrencyCodes["USS"] = "uss";
    CurrencyCodes["UYI"] = "uyi";
    CurrencyCodes["UYU"] = "uyu";
    CurrencyCodes["UZS"] = "uzs";
    CurrencyCodes["VEF"] = "vef";
    CurrencyCodes["VND"] = "vnd";
    CurrencyCodes["VUV"] = "vuv";
    CurrencyCodes["WST"] = "wst";
    CurrencyCodes["XAF"] = "xaf";
    CurrencyCodes["XAG"] = "xag";
    CurrencyCodes["XAU"] = "xau";
    CurrencyCodes["XBA"] = "xba";
    CurrencyCodes["XBB"] = "xbb";
    CurrencyCodes["XBC"] = "xbc";
    CurrencyCodes["XBD"] = "xbd";
    CurrencyCodes["XCD"] = "xcd";
    CurrencyCodes["XDR"] = "xdr";
    CurrencyCodes["XFU"] = "xfu";
    CurrencyCodes["XOF"] = "xof";
    CurrencyCodes["XPD"] = "xpd";
    CurrencyCodes["XPF"] = "xpf";
    CurrencyCodes["XPT"] = "xpt";
    CurrencyCodes["XSU"] = "xsu";
    CurrencyCodes["XTS"] = "xts";
    CurrencyCodes["XUA"] = "xua";
    CurrencyCodes["YER"] = "yer";
    CurrencyCodes["ZAR"] = "zar";
    CurrencyCodes["ZMW"] = "zmw";
    CurrencyCodes["ZWL"] = "zwl";
})(exports.CurrencyCodes || (exports.CurrencyCodes = {}));
const CurrencyExponents = {
    [exports.CurrencyCodes.AED]: 2,
    [exports.CurrencyCodes.AFN]: 2,
    [exports.CurrencyCodes.ALL]: 2,
    [exports.CurrencyCodes.AMD]: 2,
    [exports.CurrencyCodes.ANG]: 2,
    [exports.CurrencyCodes.AOA]: 2,
    [exports.CurrencyCodes.ARS]: 2,
    [exports.CurrencyCodes.AUD]: 2,
    [exports.CurrencyCodes.AWG]: 2,
    [exports.CurrencyCodes.AZN]: 2,
    [exports.CurrencyCodes.BAM]: 2,
    [exports.CurrencyCodes.BBD]: 2,
    [exports.CurrencyCodes.BDT]: 2,
    [exports.CurrencyCodes.BGN]: 2,
    [exports.CurrencyCodes.BHD]: 3,
    [exports.CurrencyCodes.BIF]: 0,
    [exports.CurrencyCodes.BMD]: 2,
    [exports.CurrencyCodes.BND]: 2,
    [exports.CurrencyCodes.BOB]: 2,
    [exports.CurrencyCodes.BOV]: 2,
    [exports.CurrencyCodes.BRL]: 2,
    [exports.CurrencyCodes.BSD]: 2,
    [exports.CurrencyCodes.BTN]: 2,
    [exports.CurrencyCodes.BWP]: 2,
    [exports.CurrencyCodes.BYR]: 0,
    [exports.CurrencyCodes.BYN]: 2,
    [exports.CurrencyCodes.BZD]: 2,
    [exports.CurrencyCodes.CAD]: 2,
    [exports.CurrencyCodes.CDF]: 2,
    [exports.CurrencyCodes.CHE]: 2,
    [exports.CurrencyCodes.CHF]: 2,
    [exports.CurrencyCodes.CHW]: 2,
    [exports.CurrencyCodes.CLF]: 0,
    [exports.CurrencyCodes.CLP]: 0,
    [exports.CurrencyCodes.CNY]: 2,
    [exports.CurrencyCodes.COP]: 2,
    [exports.CurrencyCodes.COU]: 2,
    [exports.CurrencyCodes.CRC]: 2,
    [exports.CurrencyCodes.CUC]: 2,
    [exports.CurrencyCodes.CUP]: 2,
    [exports.CurrencyCodes.CVE]: 2,
    [exports.CurrencyCodes.CZK]: 2,
    [exports.CurrencyCodes.DJF]: 0,
    [exports.CurrencyCodes.DKK]: 2,
    [exports.CurrencyCodes.DOP]: 2,
    [exports.CurrencyCodes.DZD]: 2,
    [exports.CurrencyCodes.EGP]: 2,
    [exports.CurrencyCodes.ERN]: 2,
    [exports.CurrencyCodes.ETB]: 2,
    [exports.CurrencyCodes.EUR]: 2,
    [exports.CurrencyCodes.FJD]: 2,
    [exports.CurrencyCodes.FKP]: 2,
    [exports.CurrencyCodes.GBP]: 2,
    [exports.CurrencyCodes.GEL]: 2,
    [exports.CurrencyCodes.GHS]: 2,
    [exports.CurrencyCodes.GIP]: 2,
    [exports.CurrencyCodes.GMD]: 2,
    [exports.CurrencyCodes.GNF]: 0,
    [exports.CurrencyCodes.GTQ]: 2,
    [exports.CurrencyCodes.GYD]: 2,
    [exports.CurrencyCodes.HKD]: 2,
    [exports.CurrencyCodes.HNL]: 2,
    [exports.CurrencyCodes.HRK]: 2,
    [exports.CurrencyCodes.HTG]: 2,
    [exports.CurrencyCodes.HUF]: 2,
    [exports.CurrencyCodes.IDR]: 2,
    [exports.CurrencyCodes.ILS]: 2,
    [exports.CurrencyCodes.INR]: 2,
    [exports.CurrencyCodes.IQD]: 3,
    [exports.CurrencyCodes.IRR]: 2,
    [exports.CurrencyCodes.ISK]: 0,
    [exports.CurrencyCodes.JMD]: 2,
    [exports.CurrencyCodes.JOD]: 3,
    [exports.CurrencyCodes.JPY]: 0,
    [exports.CurrencyCodes.KES]: 2,
    [exports.CurrencyCodes.KGS]: 2,
    [exports.CurrencyCodes.KHR]: 2,
    [exports.CurrencyCodes.KMF]: 0,
    [exports.CurrencyCodes.KPW]: 2,
    [exports.CurrencyCodes.KRW]: 0,
    [exports.CurrencyCodes.KWD]: 3,
    [exports.CurrencyCodes.KYD]: 2,
    [exports.CurrencyCodes.KZT]: 2,
    [exports.CurrencyCodes.LAK]: 2,
    [exports.CurrencyCodes.LBP]: 2,
    [exports.CurrencyCodes.LKR]: 2,
    [exports.CurrencyCodes.LRD]: 2,
    [exports.CurrencyCodes.LSL]: 2,
    [exports.CurrencyCodes.LTL]: 2,
    [exports.CurrencyCodes.LVL]: 2,
    [exports.CurrencyCodes.LYD]: 3,
    [exports.CurrencyCodes.MAD]: 2,
    [exports.CurrencyCodes.MDL]: 2,
    [exports.CurrencyCodes.MGA]: 2,
    [exports.CurrencyCodes.MKD]: 2,
    [exports.CurrencyCodes.MMK]: 2,
    [exports.CurrencyCodes.MNT]: 2,
    [exports.CurrencyCodes.MOP]: 2,
    [exports.CurrencyCodes.MRO]: 2,
    [exports.CurrencyCodes.MUR]: 2,
    [exports.CurrencyCodes.MVR]: 2,
    [exports.CurrencyCodes.MWK]: 2,
    [exports.CurrencyCodes.MXN]: 2,
    [exports.CurrencyCodes.MXV]: 2,
    [exports.CurrencyCodes.MYR]: 2,
    [exports.CurrencyCodes.MZN]: 2,
    [exports.CurrencyCodes.NAD]: 2,
    [exports.CurrencyCodes.NGN]: 2,
    [exports.CurrencyCodes.NIO]: 2,
    [exports.CurrencyCodes.NOK]: 2,
    [exports.CurrencyCodes.NPR]: 2,
    [exports.CurrencyCodes.NZD]: 2,
    [exports.CurrencyCodes.OMR]: 3,
    [exports.CurrencyCodes.PAB]: 2,
    [exports.CurrencyCodes.PEN]: 2,
    [exports.CurrencyCodes.PGK]: 2,
    [exports.CurrencyCodes.PHP]: 2,
    [exports.CurrencyCodes.PKR]: 2,
    [exports.CurrencyCodes.PLN]: 2,
    [exports.CurrencyCodes.PYG]: 0,
    [exports.CurrencyCodes.QAR]: 2,
    [exports.CurrencyCodes.RON]: 2,
    [exports.CurrencyCodes.RSD]: 2,
    [exports.CurrencyCodes.RUB]: 2,
    [exports.CurrencyCodes.RWF]: 0,
    [exports.CurrencyCodes.SAR]: 2,
    [exports.CurrencyCodes.SBD]: 2,
    [exports.CurrencyCodes.SCR]: 2,
    [exports.CurrencyCodes.SDG]: 2,
    [exports.CurrencyCodes.SEK]: 2,
    [exports.CurrencyCodes.SGD]: 2,
    [exports.CurrencyCodes.SHP]: 2,
    [exports.CurrencyCodes.SLL]: 2,
    [exports.CurrencyCodes.SOS]: 2,
    [exports.CurrencyCodes.SRD]: 2,
    [exports.CurrencyCodes.SSP]: 2,
    [exports.CurrencyCodes.STD]: 2,
    [exports.CurrencyCodes.SVC]: 2,
    [exports.CurrencyCodes.SYP]: 2,
    [exports.CurrencyCodes.SZL]: 2,
    [exports.CurrencyCodes.THB]: 2,
    [exports.CurrencyCodes.TJS]: 2,
    [exports.CurrencyCodes.TMT]: 2,
    [exports.CurrencyCodes.TND]: 3,
    [exports.CurrencyCodes.TOP]: 2,
    [exports.CurrencyCodes.TRY]: 2,
    [exports.CurrencyCodes.TTD]: 2,
    [exports.CurrencyCodes.TWD]: 2,
    [exports.CurrencyCodes.TZS]: 2,
    [exports.CurrencyCodes.UAH]: 2,
    [exports.CurrencyCodes.UGX]: 0,
    [exports.CurrencyCodes.USD]: 2,
    [exports.CurrencyCodes.USN]: 2,
    [exports.CurrencyCodes.USS]: 2,
    [exports.CurrencyCodes.UYI]: 0,
    [exports.CurrencyCodes.UYU]: 2,
    [exports.CurrencyCodes.UZS]: 2,
    [exports.CurrencyCodes.VEF]: 2,
    [exports.CurrencyCodes.VND]: 0,
    [exports.CurrencyCodes.VUV]: 0,
    [exports.CurrencyCodes.WST]: 2,
    [exports.CurrencyCodes.XAF]: 0,
    [exports.CurrencyCodes.XAG]: 0,
    [exports.CurrencyCodes.XAU]: 0,
    [exports.CurrencyCodes.XBA]: 0,
    [exports.CurrencyCodes.XBB]: 0,
    [exports.CurrencyCodes.XBC]: 0,
    [exports.CurrencyCodes.XBD]: 0,
    [exports.CurrencyCodes.XCD]: 2,
    [exports.CurrencyCodes.XDR]: 0,
    [exports.CurrencyCodes.XFU]: 0,
    [exports.CurrencyCodes.XOF]: 0,
    [exports.CurrencyCodes.XPD]: 0,
    [exports.CurrencyCodes.XPF]: 0,
    [exports.CurrencyCodes.XPT]: 0,
    [exports.CurrencyCodes.XSU]: 0,
    [exports.CurrencyCodes.XTS]: 0,
    [exports.CurrencyCodes.XUA]: 0,
    [exports.CurrencyCodes.YER]: 2,
    [exports.CurrencyCodes.ZAR]: 2,
    [exports.CurrencyCodes.ZMW]: 2,
    [exports.CurrencyCodes.ZWL]: 2,
};

exports.CurrencyExponents = CurrencyExponents;
